### PR TITLE
[WIP] Fixed #25605 -- Made GeoDjango DB functions work with expressions.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/25605

### CODE REVIEW

File: tests/gis_tests/distapp/tests.py [hunk 1]
```diff
@@ -517,9 +528,7 @@ def test_distance_projected(self):
             dist_qs = [dist1]
         else:
             dist2 = SouthTexasCityFt.objects.annotate(distance=Distance('point', lagrange)).order_by('id')
-            # Using EWKT string parameter.
-            dist3 = SouthTexasCityFt.objects.annotate(distance=Distance('point', lagrange.ewkt)).order_by('id')
```
Comment (claudep) [inline tests/gis_tests/distapp/tests.py]: I'd like to know why this part of the test has been dropped. If the patch is dropping a functionality, it should be documented.

File: django/contrib/gis/db/models/functions.py [hunk 2]
```diff
@@ -89,20 +90,6 @@ def as_oracle(self, compiler, connection):
...
-        if not hasattr(geom, 'srid'):
-            # Try to interpret it as a geometry input
-            try:
-                geom = GEOSGeometry(geom)
-            except Exception:
-                raise ValueError("This function requires a geometric parameter.")
-        if not geom.srid:
-            raise ValueError("Please provide a geometry attribute with a defined SRID.")
-        geom = GeomValue(geom)
-        super(GeoFuncWithGeoParam, self).__init__(expression, geom, *expressions, **extra)
```
Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: This class was doing some checks about the geometric parameter. It's clear that if we want to accept more widely expressions, it will not be possible to check that like we did it before. However, I'd like to see new tests about what happens when wrong parameter (not geometric or without srids) are passed to geometric functions. We should ensure that the error messages are clear for the developer.

File: tests/gis_tests/distapp/tests.py (hunk 1 above)
Comment (sir-sigurd) [inline tests/gis_tests/distapp/tests.py]: I fixed calling with input for GEOSGeometry in update, though IMO this is overkill.

File: django/contrib/gis/db/models/functions.py [hunk 3]
```diff
@@ -125,6 +161,9 @@ def as_oracle(self, compiler, connection):
 
 
 class Area(OracleToleranceMixin, GeoFunc):
+    def __init__(self, expression, **extra):
+        super(Area, self).__init__(expression, **extra)
```
Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: You added several `__init__` methods with no apparent role. Why is that?

Comment (sir-sigurd) [inline django/contrib/gis/db/models/functions.py]: to check arity

Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: Sorry, could you explain what it's supposed to change in practice?

Comment (sir-sigurd) [inline django/contrib/gis/db/models/functions.py]: before
```
In [1]: from django.contrib.gis.db.models.functions import Area

In [2]: from geotest.models import Test          

In [3]: Test.objects.annotate(area=Area('point1', 'point2'))
Out[3]: ---------------------------------------------------------------------------
ProgrammingError                          Traceback (most recent call last)
...
ProgrammingError: ОШИБКА:  функция st_area(geometry, geometry) не существует
```

after
```
In [1]: from django.contrib.gis.db.models.functions import Area

In [2]: from geotest.models import Test

In [3]: Test.objects.annotate(area=Area('point1', 'point2'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-798e1a5dd312> in <module>()
----> 1 Test.objects.annotate(area=Area('point1', 'point2'))

TypeError: __init__() takes exactly 2 arguments (3 given)
```

Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: Many thanks for the example. I guess this is the same before the patch, is it? Then we might fix that in a separate patch. That would keep this patch a bit smaller.

Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: I've created https://code.djangoproject.com/ticket/25629 to track this.

Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: Regarding the implementation, I'd rather have some class variable (`max_expressions`?) checked in base class instead of "dumb" `__init__` methods. To be experimented...

File: tests/gis_tests/distapp/tests.py (hunk 1 above)
Comment (sir-sigurd) [inline tests/gis_tests/distapp/tests.py]: https://docs.djangoproject.com/en/1.8/ref/models/expressions/#django.db.models.Func

> Positional arguments can be expressions or Python values. **Strings are assumed to be column references and will be wrapped in F() expressions** while other values will be wrapped in Value() expressions.

Do you still think we should handle geo_input parameters?

Comment (claudep) [inline tests/gis_tests/distapp/tests.py]: Note that I didn't plea for keeping the functionality. I just said that if we remove that, it should be done properly (backwards incompatibility note, separate commit).

Comment (sir-sigurd) [inline tests/gis_tests/distapp/tests.py]: Ah, ok. Well I don't think this behavior is documented anywhere.

Comment (sir-sigurd) [inline tests/gis_tests/distapp/tests.py]: Are we targeting this for 1.9 or for 1.10? If for 1.10 then we should fix documentation for 1.9 making clear what arguments are expected.

Comment (claudep) [inline tests/gis_tests/distapp/tests.py]: We are mostly targeting 1.10, and if some parts need backporting to 1.9, just suggest it on the PR.

**Comment by claudep** [comment]
Could you please rebase this one?

**Comment by sir-sigurd** [comment]
I'm still working on it, probably will push update tomorrow.

**Comment by claudep** [comment]
I worked a bit on this issue this morning and here's the result: claudep@69e2fb85b9c2d74c0cadc38aad2bfe4b9ccf8cda
It's not yet thoroughly tested (as yours) but what do you think about the approach taken?

**Comment by sir-sigurd** [comment]
Well, I'm still sure that we should remove handling of string that differs from [`Func`](https://docs.djangoproject.com/en/1.8/ref/models/expressions/#func-expressions), because WKT parameters could mask fields (this example was done on your branch):

```
In [33]: qs = Test.objects.annotate(**{'SRID=4326;POINT (10 10)': Centroid('point')})

In [34]: qs.values().first()
Out[34]: 
{'SRID=4326;POINT (10 10)': <Point object at 0x7f094f0827f0>,
 u'id': 1,
 'name': u'name',
 'point': <Point object at 0x7f094f082880>}

In [35]: qs = qs.annotate(union=Union('SRID=4326;POINT (10 10)', 'point'))

In [36]: {k: str(v) for k, v in qs.values().first().items()}
Out[36]: 
{'SRID=4326;POINT (10 10)': 'SRID=4326;POINT (0.0000000000000000 0.0000000000000000)',
 u'id': '1',
 'name': 'name',
 'point': 'SRID=4326;POINT (0.0000000000000000 0.0000000000000000)',
 'union': 'SRID=4326;MULTIPOINT (10.0000000000000000 10.0000000000000000, 0.0000000000000000 0.0000000000000000)'}
```
Though this example is speculative (but still valid), there is no guarantee that `GEOSGeometry` will not support some formats more similar to common field names.

**Comment by claudep** [comment]
Makes sense. Still undecided about adding deprecation or not for this removal...

**Comment by sir-sigurd** [comment]
I think we should just remove it from 1.9 while it's not released.

**Comment by sir-sigurd** [comment]
I mean only new GIS functions, not old GeoManager methods.

**Comment by claudep** [comment]
Tested that in #5583. If tests are happy and you approve it, I'll commit and backport.

**Comment by claudep** [comment]
WIP patch updated in claudep@304d700057

File: django/contrib/gis/db/models/functions.py [hunk 4]
```diff
@@ -408,14 +451,7 @@ def __init__(self, expression, srid, **extra):
...
     def srid(self):
         # Make srid the resulting srid of the transformation
-        return self.source_expressions[self.geom_param_pos + 1].value
-
-    def convert_value(self, value, expression, connection, context):
-        value = super(Transform, self).convert_value(value, expression, connection, context)
-        if not connection.ops.postgis and not value.srid:
-            # Some backends do not set the srid on the returning geometry
-            value.srid = self.srid
-        return value
```
Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: Is that removal related to 5f7035cec7d5? If yes, we might commit it separately with an appropriate message.

File: django/contrib/gis/db/models/functions.py [hunk 5]
```diff
@@ -72,10 +114,7 @@ def srid(self):
         return self.value.srid
 
     def as_sql(self, compiler, connection):
-        if self.geography:
-            self.value = connection.ops.Adapter(self.value, geography=self.geography)
-        else:
-            self.value = connection.ops.Adapter(self.value)
+        self.value = connection.ops.Adapter(self.value, geography=self.geography)
```
Comment (claudep) [inline django/contrib/gis/db/models/functions.py]: Is that simplification related to anything else in this patch?

**Comment by claudep** [comment]
The title of the PR stills contains WIP, is it really still WIP?

**Comment by sir-sigurd** [comment]
Yes, this is still WIP, I think we should make decision on [this](https://github.com/django/django/pull/5614) at first.

**Comment by claudep** [comment]
OK, I'll push that one on my review list first.

**Comment by timgraham** [comment]
Closing due to inactivity.
